packages/cli/src/api/formats/po.js

```diff
@@ -1,104 +1,113 @@
+    item.obsolete = message.obsolete
+    item.flags = message.flags
+      ? R.fromPairs(message.flags.map(flag => [flag, true]))
+      : {}
+    return item
   })
 )
 
-function parseOrigin(origins) {
-  if (!origins) return
+const getMessageKey = R.prop("msgid")
+const getTranslations = R.prop("msgstr")
+const getExtractedComments = R.prop("extractedComments")
+const getTranslatorComments = R.prop("comments")
+const getOrigins = R.prop("references")
+const getFlags = R.compose(
+  R.map(R.trim),
+  R.keys,
+  R.dissoc("obsolete"), // backward-compatibility, remove in 3.x
+  R.prop("flags")
+)
+const isObsolete = R.either(R.path(["flags", "obsolete"]), R.prop("obsolete"))
+
+const deserialize = R.map(
+  R.applySpec({
+    translation: R.compose(
+      R.head,
+      R.defaultTo([]),
+      getTranslations
+    ),
+    description: R.compose(
+      R.head,
+      R.defaultTo([]),
+      getExtractedComments
+    ),
+    comments: item =>
+      R.concat(getTranslatorComments(item), R.tail(getExtractedComments(item))),
+    obsolete: isObsolete,
+    origin: R.compose(
+      R.map(splitOrigin),
+      R.defaultTo([]),
+      getOrigins
+    ),
+    flags: getFlags
+  })
+)
 
-  // split by comma or newlines, but don't remember matched delimiter
-  return origins.split(/(?:, |\n)/).map(origin => origin.split(":"))
-}
+const validateItems = R.map(item => {
+  if (R.length(getTranslations(item)) > 1) {
+    console.warn(
+      "Multiple translations for item with key %s is not supported and it will be ignored.",
+      getMessageKey(item)
+    )
+  }
+})
 
-const deserialize = R.map(message => ({
-  translation: message.msgstr.toString(),
-  description: R.path(["comments", "extracted"], message),
-  obsolete: R.pathEq(["comments", "flag"], "obsolete", message),
-  origin: parseOrigin(R.path(["comments", "reference"], message))
-}))
+const indexItems = R.indexBy(getMessageKey)
 
 const format: TranslationsFormat = {
   filename: "messages.po",
 
   write(filename, catalog, options = {}) {
-    let headers
+    let po
+    let indexedItems = {}
     if (fs.existsSync(filename)) {
       const raw = fs.readFileSync(filename).toString()
-      const prevPo = gettextParser.po.parse(raw)
-      headers = prevPo.headers
+      po = PO.parse(raw)
     } else {
-      headers = getHeaders(options)
+      po = new PO()
+      po.headers = getCreateHeaders(options.language)
+      po.headerOrder = R.keys(po.headers)
     }
-
-    const pofile = [...writeHeaders(headers), ...serialize(catalog)].join("\n")
-    fs.writeFileSync(filename, pofile + "\n")
+    po.items = serialize(catalog)
+    fs.writeFileSync(filename, po.toString())
   },
 
   read(filename) {
     const raw = fs.readFileSync(filename).toString()
+    return this.parse(raw)
+  },
 
-    const pofile = gettextParser.po.parse(raw)
-    const translations = pofile.translations[""]
-    delete translations[""] // remove headers
-    return deserialize(translations)
+  parse(raw) {
```

Comment:
What's the reason for having separate `this.parse` method, which is used just once?

Reply:
I am trying with Loco to automate the translation process by uploading/downloading translation files. That means I need to be able to parse received PO file to catalog and I don't want to save to file system because of that. Then I can call catalog.merge and be happy hopefully :)

Reply:
Fair enough!

I used to have `sync` command to upload/download messages from 3rd party services, but I never released it. I guess everything could be handled in `extract` command. Maybe we could provide hooks, to intercept extracting/collecting messages... Please keep me posted what you manage to hack together :)